### ISSUE TITLE
major differences between ANGSD/realSFS versions

### ISSUE THREAD

**Issue by claudiuskerth** [ANGSD/angsd#97]
Hi,

I have noticed that spectra estimated from the same data and with the same command line options but either with an old (0.915-5-ge6e63e5, Nov 2016) or recent version of ANGSD/realSFS (0.917-142-ge3dbeaa) look quite different.

[compare angsd versions IPython notebook](https://www.dropbox.com/s/slp601b37w42loi/compare_angsd_versions.html?dl=0)

Also note that the number of sites in the SAF files is slightly different between the two versions.

Could you please comment on this? Is this surprising to you? Did you make major changes to the way SAF likelihoods and spectra are estimated since last November? Do you recommend that users re-estimate their SAF files, spectra, etc.?

many thanks for your help!

claudius

**Comment by claudiuskerth** [comment]
Here are the full command lines:

#### SAF files with old version of ANGSD

> \# PER POP:
> angsd -bam PAR.slim.bamfile.list -ref Big_Data_ref.fa -anc Big_Data_ref.fa -out SAFs/PAR/PAR.unfolded -fold 0 -sites $keep.sites -rf $keep.rf -only_proper_pairs 0 -baq 1 -minMapQ 5 -minInd 9 -GL 1 -doSaf 1 -nThreads 1

This calculates SAF's for 1,214,939 sites. 

> angsd -bam ERY.slim.bamfile.list -ref Big_Data_ref.fa -anc Big_Data_ref.fa -out SAFs/ERY/ERY.unfolded -fold 0 -sites $keep.sites -rf $keep.rf -only_proper_pairs 0 -baq 1 -minMapQ 5 -minInd 9 -GL 1 -doSaf 1 -nThreads 1

This calculates SAF's for 1,638,468 sites.

#### estimate ML unfolded 2D-SFS wil old realSFS

> realSFS -P 12 SAFs/ERY/ERY.unfolded.saf.idx SAFs/PAR/PAR.unfolded.saf.idx 2> /dev/null > FST/EryPar.unfolded.2dsfs

#### SAF files with new version of ANGSD

>mkdir -p SAFs/with_ANGSD-0.917-142-ge3dbeaa/ERY
> mkdir -p SAFs/with_ANGSD-0.917-142-ge3dbeaa/PAR
> split -l 500 $keep.rf split_rf/
> \# PER POP:
> ls split_rf/* | parallel -j 24 "angsd -bam PAR.slim.bamfile.list -ref Big_Data_ref.fa -anc Big_Data_ref.fa -out SAFs/with_ANGSD-0.917-142-ge3dbeaa/PAR/{/}.unfolded -fold 0 \
-sites $keep.sites -rf {} -only_proper_pairs 0 -baq 1 -minMapQ 5 -minInd 9 -GL 1 -doSaf 1 -nThreads 1 2>/dev/null"
cd SAFs/with_ANGSD-0.917-142-ge3dbeaa/PAR
realSFS cat -outnames PAR.unfolded *saf.idx

This calculates SAF's for 1,217,242 sites. This is 2,303 more sites than in the SAF file calculated with a previous version of ANGSD from exactly the same data! 

> ls split_rf/* | parallel -j 24 "angsd -bam ERY.slim.bamfile.list -ref Big_Data_ref.fa -anc Big_Data_ref.fa  -out SAFs/with_ANGSD-0.917-142-ge3dbeaa/ERY/{/}.unfolded -fold 0 -sites $keep.sites -rf {} -only_proper_pairs 0 -baq 1 -minMapQ 5 -minInd 9 -GL 1 -doSaf 1 -nThreads 1 2>/dev/null"
> cd SAFs/with_ANGSD-0.917-142-ge3dbeaa/ERY
> realSFS cat -outnames ERY.unfolded *saf.idx

This calculates SAF's for 1,639,873 sites. This is 1,405 more sites than in the SAF file calculated with a previous version of ANGSD from exactly the same data!

#### estimate ML unfolded 2D-SFS with new realSFS:

> mkdir -p SFS/with_ANGSD-0.917-142-ge3dbeaa/EryPar
> realSFS -P 24 SAFs/with_ANGSD-0.917-142-ge3dbeaa/ERY/ERY.unfolded.saf.idx SAFs/with_ANGSD-0.917-142-ge3dbeaa/PAR/PAR.unfolded.saf.idx \
2>/dev/null > SFS/with_ANGSD-0.917-142-ge3dbeaa/EryPar/EryPar.unfolded.2dsfs

Obviously, the same "keep" file was used throughout. The 2D-SFS from the new version of ANGSD contains 2,748 more sites (including monomorphic).

claudius

**Comment by claudiuskerth** [comment]
The difference between 2D-SFS's from old and new SAF files is not due to insufficient convergence during the estimation of one or the other. I have re-estimated the 2D-SFS from old and new SAF files with exhaustive search parameters for realSFS: `-maxIter 50000 -tole 1e-6 -m 0`. This does change the 2D-SFS's only to a minor degree and the difference between them is hardly changed.

Furthermore a diff between old and new SAF files confirms that SAF likelihood calculation has changed since November 2016 ([diff between SAF files](https://www.dropbox.com/s/e24jhhd9s6u70o2/compare_PAR_old_new_SAFs.tab?dl=0)). Just by looking at the general shape of the marginal spectra of my 2D-SFS's, I would say that SAF likelihood calculation has improved. 

Given the size of the difference in the spectra, I suspect that results from downstream analyses could change at least quantitatively depending on whether they were done with angsd version 0.915-5-ge6e63e5 or version 0.917-142-ge3dbeaa. 

claudius

**Comment by ANGSD** [comment]
I dont think there has been any changes in the core saf calculation but there has been some changes in the baq calculation some months ago. I think the default in samtools is to do extended baq. Which correspond to -baq 2 in  the newer angsd. standard baq is -baq 1. Hopefully this explains the difference. One other thing to consider is whether or not to use baq at all. The newer versions of gatk has removed baq all together. 

Best

**Comment by claudiuskerth** [comment]
Hi Thorfinn,

the _recent_ version of angsd/realSFS (0.917-142-ge3dbeaa, latest commit e3dbeaaef06) that I used has the commit 6dbc9284 from the 31 of May with the message:

> Changed parsing of -baq 1 to match default(extended baq) in SAMtools. -baq 2 (redo baq)

This is the most recent commit mentioning BAQ in this version. 

I also see that in the commit from the 22nd of June 3367eff, which is more recent than the version I used, the meaning of `-baq` arguments changed again:

> only supported options for -baq is: 1 (normal baq) and 2 (extended baq (SAMtools default))

So, did I use extended BAQ with my _recent_ version from commit e3dbeaaef06? Did angsd/realSFS versions from _before_ the commit 6dbc9284 from the 31 of May use standard BAQ?

claudius

**Comment by claudiuskerth** [comment]
Hi,

I agree that the changes in BAQ calculation are the likely cause for the differences in the SAF's. 

Standard BAQ is probably too conservative ([samtools mailing list](https://sourceforge.net/p/samtools/mailman/message/28933512/)) and therefore should reduce diversity estimates and maybe even bias the SFS. Not applying BAQ is probably not a good idea either: [samtools mailing list](https://sourceforge.net/p/samtools/mailman/message/27766885/) and

>Currently the BAQ strategy is the only practical way to avoid the INDEL artifact in low-coverage multi-sample SNP calling.

http://samtools.sourceforge.net/mpileup.shtml

So why did GATK remove BAQ adjustment altogether? I believe this is because HaplotypeCaller now does _local re-assembly of reads_ in variable regions, which seems to be the best way to identify and align reads around indels ([HaplotypeCaller](https://software.broadinstitute.org/gatk/documentation/article?id=4148)). _Indel realignment_, on the other hand, does not always seem to do a good job ([Li2014, section 3.2.2](https://academic.oup.com/bioinformatics/article/30/20/2843/2422145/Toward-better-understanding-of-artifacts-in)). I did indel realignment with GATK, but it did not change the alignment of any read. Removing sites from low complexity regions (e. g. with dustmasker) should alleviate to some extent the problem of false positive SNP's around small indels.

claudius

**Comment by ANGSD** [comment]
From looking at the samtools code and some small test it looked like
samtools only does extended baq and not normal baq. The older version of
angsd did normal baq with -dobaq 1. It is possible to do extended baq with
-dobaq 2. These are always recalculated and are not using any bq tags if
they exist.

**Comment by claudiuskerth** [comment]
A little follow-up:

Although @ANGSD has not answered my questions above unambiguously about which versions of BAQ I used when using specific versions of ANGSD (with the same command lines!), I am assuming that I used standard BAQ with the old version (Nov 2016) and extended BAQ with a version from June 2017. 

Previously I wrote:

>Given the size of the difference in the spectra, I suspect that results from downstream analyses could change at least quantitatively depending on whether they were done with angsd version 0.915-5-ge6e63e5 or version 0.917-142-ge3dbeaa.

I can now say that as far as demographic inference with dadi is concerned the difference is quantitative in that differences between ML parameter estimates are sometimes large, but the same models come up as most likely. In contrast, stairway-plot (from 1D SFS) infers a different population size history for one of the two populations depending on whether standard BAQ or extended BAQ was used. For the other population, the version of BAQ had only a minor influence on inferred population size history.

Bottom line: BAQ _can_ influence your results and sometimes even conclusions! 

claudius

**Comment by ANGSD** [comment]
hello claudius, sorry for not being more response I've been away and are
slowly getting back on track.

From the back of my head the oldest version with -dobaq 1 would correspond
to standard baq. Then in some intermediate version -dobaq 1 would be
extended baq. Then in the final iteration it was reverted back such that
-dobaq 1 is standard baq and -dobaq 2 is extended back. I dont think it is
possible to do anything but extended baq in the samtools mpileup routine
but i could be wrong.

I agree that baq will have an impact. But i dont think it is obvious which
level of baq is the proper/correct, it might be related to the difference
and quality of the reference genome. I got feedback from another user
working with adna that extended baq would cause too many hets.
